src/object/cli_obj.c

```diff
@@ -559,11 +559,21 @@ obj_dkey2grpmemb(struct dc_object *obj, uint64_t hash, uint32_t map_ver,
 	return 0;
 }
 
-static uint32_t
-obj_shard2tgtid(struct dc_object *obj, uint32_t shard)
+static int
+obj_shard2tgtid(struct dc_object *obj, uint32_t shard, uint32_t map_ver,
+		uint32_t *tgt_id)
 {
-	D_ASSERT(shard < obj->cob_shards_nr);
-	return obj->cob_shards->do_shards[shard].do_target_id;
+	D_RWLOCK_RDLOCK(&obj->cob_lock);
+	if (map_ver == obj->cob_version)
+		D_ASSERT(shard < obj->cob_shards_nr);
+	if (shard >= obj->cob_shards_nr) {
+		D_RWLOCK_UNLOCK(&obj->cob_lock);
+		return -DER_NONEXIST;
```

Comment:
It looks to me that it could fail fetch to an object with redundancy (when a redundancy degree is lost), for example, a 2-way replica, if one target is DOWN and no available spare target is available, the shard count of object layout will be decreased to 1 (less the the original 2), will the fetch failed for -DER_NONEXIST? Or there is any code in caller code path which tolerating this -DER_NONEXIST error for fetch?

Reply:
The obj->cob_shards_nr is set as  layout->ol_nr, that is always set as grp_nr * grp_size.
so the code change here should only affect the case of losing whole RDG group when grp_nr = GRP_MAX.

In the case you raised, 2-way replica lost one tgt, grp_nr * grp_size is not changed, but the do_target_id[lost_shard] should be set as -1.

Reply:
Got it, thanks!